src/LinuxTracing/PerfEventQueue.cpp

```diff
@@ -25,41 +29,66 @@ void PerfEventQueue::PushEvent(std::unique_ptr<PerfEvent> event) {
     queue->push(std::move(event));
 
   } else {
-    queue_it =
-        queues_.emplace(origin_fd, std::make_unique<std::queue<std::unique_ptr<PerfEvent>>>())
-            .first;
+    queue_it = ordered_queues_by_fd_
+                   .emplace(origin_fd, std::make_unique<std::queue<std::unique_ptr<PerfEvent>>>())
+                   .first;
     const std::unique_ptr<std::queue<std::unique_ptr<PerfEvent>>>& queue = queue_it->second;
 
     queue->push(std::move(event));
-    queues_heap_.emplace_back(queue.get());
-    MoveUpHeapBack();
+    ordered_queues_heap_.emplace_back(queue.get());
+    MoveUpBackOfOrderedQueuesHeap();
   }
 }
 
-bool PerfEventQueue::HasEvent() { return !queues_heap_.empty(); }
+bool PerfEventQueue::HasEvent() const {
+  return !ordered_queues_heap_.empty() || !unordered_events_priority_queue_.empty();
+}
 
-PerfEvent* PerfEventQueue::TopEvent() { return queues_heap_.front()->front().get(); }
+PerfEvent* PerfEventQueue::TopEvent() {
+  PerfEvent* top_event = nullptr;
+  if (!unordered_events_priority_queue_.empty()) {
```

Comment:
I don't quite get what is happening here.
If we have an unordered event at the top that has a timestamp greater then the top of the ordered events, it doesn't mean, that there isn't another unordered event with a smaller timestamp. Is that important?

Reply:
Hmm, I don't really understand what you didn't understand. Potentially we have two "top events" (from the two queues) here and we want to get the older of the two. I added a comment with that by the way.

Reply:
Thanks for the comment. My point is, that at a first glance, this code really looks like you want to get THE smallest event (by timestamp). But this just not the case, as there might be smaller elements in the unordered_events. However, as this seems to be expected, I am probably fine.

Reply:
Yes, I want to get the smallest event by timestamp. `unordered_events_priority_queue_` is a priority queue and is sorted by timestamp. Or what am I missing?

Reply:
Discussed offline, added a commit that makes field names clearer.